### PR TITLE
[SPARK-31066][SQL][test-hive1.2] Disable useless and uncleaned hive SessionState initialization parts

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

As a common usage and according to the spark doc, users may often just copy their `hive-site.xml` to Spark directly from hive projects. Sometimes, the config file is not that clean for spark and may cause some side effects.

for example, `hive.session.history.enabled` will create a log for the hive jobs but useless for spark and also it will not be deleted on JVM exit.

this pr
 1) disable `hive.session.history.enabled` explicitly to disable creating `hive_job_log` file, e.g.
```
Hive history file=/var/folders/01/h81cs4sn3dq2dd_k4j6fhrmc0000gn/T//kentyao/hive_job_log_79c63b29-95a4-4935-a9eb-2d89844dfe4f_493861201.txt
```
2) set `hive.execution.engine` to `spark` explicitly in case the config is `tez` and casue uneccesary problem like this:

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/tez/dag/api/SessionNotRunning
	at org.apache.hadoop.hive.ql.session.SessionState.start(SessionState.java:529)
```

### Why are the changes needed?

reduce overhead of internal complexity and users' hive cognitive load for  running spark

### Does this PR introduce any user-facing change?

yes, `hive_job_log` file will not be created even enabled, and will not try to initialize tez kinds of stuff
### How was this patch tested?

add ut and verify manually

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @dongjoon-hyun, thanks for reviewing this.

**Comment by maropu** [comment]
also cc: @wangyum

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -1232,4 +1204,42 @@ private[hive] object HiveClientImpl {
...
+      logDebug(s"Applying Hadoop/Hive/Spark and extra properties to Hive Conf:$k=$v")
+    }
+    // Disable CBO because we removed the Calcite dependency.
+    hiveConf.setBoolean("hive.cbo.enable", false)
+    // If this is true, SessionState.start will create a file to log hive job which will not be
+    // deleted on exit and is useless for spark
+    hiveConf.setBoolean("hive.session.history.enabled", false)
+    // If this is tez engine, SessionState.start might bring extra logic to initialize tez stuff,
+    // which is useless for spark.
+    hiveConf.set("hive.execution.engine", "spark")
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: The overwrite is better than just a warning log?

**Comment by maropu** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: warning log is not enough, we need reset this before hive `SessionState.start`. And the reason I did not add a warning here is to follow what we did for `hive.cbo.enable`

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by maropu** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 1 above)
Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Ah, I see. But, is it ok to silently rewrite the config? The overwrite itself ok, but I feel it'd be better to inform users that they should change their config in a next run.

Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I have added warning

**Comment by yaooqinn** [comment]
gentle ping @maropu

**Comment by maropu** [comment]
Is this fix related only to `hive1.2`? I just want to know which hive version is related to this issue.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 1 above)
Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: WDYT this fix? @dongjoon-hyun @HyukjinKwon ?

**Comment by yaooqinn** [comment]
> Is this fix related only to `hive1.2`? I just want to know which hive version is related to this issue.

No, this mark is not added by me, I have no idea why it is needed

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -159,36 +160,7 @@ private[hive] class HiveClientImpl(
...
-    confMap.foreach { case (k, v) => hiveConf.set(k, v) }
-    SQLConf.get.redactOptions(confMap).foreach { case (k, v) =>
-      logDebug(
-        s"""
-           |Applying Hadoop/Hive/Spark and extra properties to Hive Conf:
-           |$k=$v
-         """.stripMargin)
-    }
-    // Disable CBO because we removed the Calcite dependency.
-    hiveConf.setBoolean("hive.cbo.enable", false)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: It's not ideal to handle hive-specific configs in Spark. But we already did here, and it makes Spark more user-friendly, I'm OK with it.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by gatorsmile** [comment]
cc @juliuszsompolski

**Pull request closed without merging**